File: internal/radio/theme_modal.go
package radio

import (
	"fmt"
	"strings"

	"github.com/gdamore/tcell/v2"
	"github.com/rivo/tview"
)

// The theme modal previews the theme under the cursor on the whole UI. Enter
// saves it, Esc goes back to the saved one.

const themeNameWidth = 16

func (a *Application) setupThemeModal() {
	a.themeList = newList()
	a.themeList.SetBorder(true).SetBorderPadding(0, 0, 1, 1).
		SetTitle(" Theme ").SetTitleAlign(tview.AlignLeft)
	a.themeList.SetHighlightFullLine(true).SetWrapAround(false)
	for i, t := range themes {
		a.themeList.AddItem(t.name, "", idxToRune(i), func() { a.saveTheme(t.name) })
	}
	a.themeList.SetChangedFunc(func(index int, _, _ string, _ rune) {
		a.setTheme(themes[index])
	})
	a.themeList.SetInputCapture(func(event *tcell.EventKey) *tcell.EventKey {
		if event.Key() == tcell.KeyEscape {
			a.cancelThemeModal()
			return nil
		}
		return event
	})

	width := 4 + 2 + themeNameWidth + 2 + 4*2 + 4 // shortcut, mark, name, gap, swatches, frame
	modal := centerIn(tview.NewFlex(), a.themeList, width, len(themes)+2)
	a.pages.AddPage(a.pageNames[ThemePage], modal, true, false)
}

// themeLabel shows the theme's own colours, so themes can be compared without
// previewing each one.
func themeLabel(t theme, saved bool) string {
	mark := "  "
	if saved {
		mark = glyphs.check + " "
	}
	var sb strings.Builder
	fmt.Fprintf(&sb, "%s%-*s  ", mark, themeNameWidth, t.name)
	for _, c := range []tcell.Color{t.accent, t.bright, t.warn, t.danger} {
		sb.WriteString(fgTag(c) + glyphs.swatch)
	}
	sb.WriteString("[-]")
	return sb.String()
}

func (a *Application) showThemeModal() {
	a.hideSearchModal()
	a.hideRemoteModal()
	current := 0
	for i, t := range themes {
		saved := t.name == a.config.Theme
		if saved {
			current = i
		}
		a.themeList.SetItemText(i, themeLabel(t, saved), "")
	}
	a.themeList.SetCurrentItem(current)
	a.pages.ShowPage(a.pageNames[ThemePage])
	a.app.SetFocus(a.themeList)
}

func (a *Application) hideThemeModal() {
	a.pages.HidePage(a.pageNames[ThemePage])
}

func (a *Application) cancelThemeModal() {
	t, _ := findTheme(a.config.Theme)
	a.setTheme(t)
	a.hideThemeModal()
}

func (a *Application) saveTheme(name string) {
	a.config.Theme = name
	if err := a.config.save(); err != nil {
		log.Printf("config: %v", err)
	}
	a.hideThemeModal()
}

func (a *Application) isThemeModalOpen() bool {
	return a.frontPage() == a.pageNames[ThemePage]
}

File: rust/src/ui/theme_modal.rs
//! The theme picker previews the theme under the cursor on the whole UI.
//! Enter saves it, Esc goes back to the saved one.

use std::io::Write;

use ratatui::buffer::Buffer;
use ratatui::crossterm::event::{KeyCode, KeyEvent, MouseButton, MouseEvent, MouseEventKind};
use ratatui::layout::{Position, Rect};
use ratatui::style::Color;

use super::list::{self, ListState, Pane, Row};
use super::text::{fill, seg};
use super::theme::{self, THEMES};
use super::{App, Modal};
use crate::log;

const NAME_WIDTH: usize = 20;

impl App {
    pub(super) fn show_theme_modal(&mut self) {
        let mut state = ListState::default();
        state.cursor = THEMES.iter().position(|t| t.name == self.config.theme).unwrap_or(0);
        self.modal = Some(Modal::Theme(state));
    }

    fn cancel_theme_modal(&mut self) {
        self.look.t = theme::find(&self.config.theme).unwrap_or(THEMES[0]);
        self.modal = None;
    }

    fn save_theme(&mut self, i: usize) {
        self.look.t = THEMES[i];
        self.config.theme = THEMES[i].name.to_string();
        if let Err(e) = self.config.save(&["theme"]) {
            log!("config: {e}");
        }
        self.modal = None;
    }

    fn theme_cursor(&mut self) -> Option<&mut ListState> {
        match &mut self.modal {
            Some(Modal::Theme(state)) => Some(state),
            _ => None,
        }
    }

    fn preview_theme(&mut self) {
        if let Some(i) = self.theme_cursor().map(|s| s.cursor) {
            self.look.t = THEMES[i];
        }
    }

    pub(super) fn on_theme_key(&mut self, k: KeyEvent) {
        let Some(state) = self.theme_cursor() else { return };
        let page = state.page();
        let delta = match k.code {
            KeyCode::Up => -1,
            KeyCode::Down => 1,
            KeyCode::PageUp => -page,
            KeyCode::PageDown => page,
            KeyCode::Home => isize::MIN / 2,
            KeyCode::End => isize::MAX / 2,
            KeyCode::Enter | KeyCode::Char(' ') => {
                let i = state.cursor;
                self.save_theme(i);
                return;
            }
            KeyCode::Esc => {
                self.cancel_theme_modal();
                return;
            }
            KeyCode::Char('t') if k.modifiers.contains(ratatui::crossterm::event::KeyModifiers::CONTROL) => {
                self.cancel_theme_modal();
                return;
            }
            _ => return,
        };
        state.move_by(delta, THEMES.len());
        self.preview_theme();
    }

    pub(super) fn on_theme_mouse(&mut self, m: MouseEvent) {
        let area = self.modal_area;
        if !area.contains(Position::new(m.column, m.row)) {
            return;
        }
        let Some(state) = self.theme_cursor() else { return };
        match m.kind {
            MouseEventKind::ScrollUp => state.scroll(-1, THEMES.len()),
            MouseEventKind::ScrollDown => state.scroll(1, THEMES.len()),
            MouseEventKind::Down(MouseButton::Left) => {
                if let Some(i) = state.row_at(m.row, THEMES.len()) {
                    self.save_theme(i);
                }
                return;
            }
            _ => return,
        }
        self.preview_theme();
    }

    /// Each row shows the theme's own colours, so themes can be compared
    /// without previewing each one.
    pub(super) fn draw_theme_modal(&mut self, buf: &mut Buffer, screen: Rect) {
        let (t, g) = (self.look.t, self.look.g);
        let rows: Vec<Row> = THEMES
            .iter()
            .map(|th| {
                let saved = th.name == self.config.theme;
                let mark = if saved { format!("{} ", g.check) } else { "  ".to_string() };
                let mut label = vec![seg(format!("{mark}{:<NAME_WIDTH$}  ", th.name), t.text())];
                for c in [th.accent, th.bright, th.warn, th.danger] {
                    label.push(seg(g.swatch, t.fg(c)));
                }
                Row { label, ..Row::default() }
            })
            .collect();
        let width = (5 + 2 + NAME_WIDTH + 2 + 4 * 2) as u16;
        let height = (THEMES.len() as u16 + 2).min(screen.height);
        let area = Rect {
            x: screen.x + screen.width.saturating_sub(width) / 2,
            y: screen.y + screen.height.saturating_sub(height) / 2,
            width: width.min(screen.width),
            height,
        };
        self.modal_area = area;
        for y in area.top()..area.bottom() {
            fill(buf, area.x, y, area.width as usize, t.text());
        }
        let pane = Pane { title: vec![seg(" Theme ", t.text())], focused: true, rows: &rows, empty: None };
        if let Some(Modal::Theme(state)) = &mut self.modal {
            list::draw(&self.look, buf, area, &pane, state);
        }
    }

    /// The terminal's padding around the cell grid can't be drawn on, so a
    /// themed background is also set as the terminal's default background
    /// (OSC 11), and reset on exit (OSC 111). Terminals without OSC 11
    /// ignore both.
    pub(super) fn sync_term_bg(&mut self) {
        let bg = self.look.t.bg;
        if bg == self.term_bg {
            return;
        }
        let seq = match bg {
            Color::Rgb(r, g, b) => format!("\x1b]11;#{r:02x}{g:02x}{b:02x}\x1b\\"),
            _ => "\x1b]111\x1b\\".to_string(),
        };
        let mut out = std::io::stdout();
        if let Err(e) = out.write_all(seq.as_bytes()).and_then(|_| out.flush()) {
            log!("terminal background: {e}");
        }
        self.term_bg = bg;
    }
}
